lib/filterx/func-str.c

```diff
@@ -0,0 +1,459 @@
+    return NULL;
+
+  return result;
+}
+
+static GString *
+_obj_format(FilterXObject *obj, gboolean ignore_case)
+{
+  GString *result = _format_str_obj(obj);
+  if (!result)
+    filterx_eval_push_error("failed to extract string value, repr() failed", NULL, obj);
+  filterx_object_unref(obj);
+
+  if (ignore_case && !_do_casefold(result))
+    return NULL;
+
+  return result;
+}
+
+static gboolean
+_string_with_cache_fill_cache(FilterXStringWithCache *self, gboolean ignore_case)
+{
+  if(!filterx_expr_is_literal(self->expr))
+    return TRUE;
+
+  return !!_expr_format(self->expr, ignore_case);
+}
+
+static FilterXStringWithCache *
+_string_with_cache_new(FilterXExpr *expr, gboolean ignore_case)
+{
+  FilterXStringWithCache *self = g_new0(FilterXStringWithCache, 1);
+  self->expr = expr;
+
+  if (!_string_with_cache_fill_cache(self, ignore_case))
+    return NULL;
+
+  return self;
+}
+
+static gboolean
+_string_with_cache_get_string_value(FilterXStringWithCache *self, gboolean ignore_case, const gchar **dst,
+                                    gsize *len)
+{
+  if(self->str_value)
+    {
+      *dst = self->str_value;
+      *len = self->str_len;
+      return TRUE;
+    }
+
+  GString *res = _expr_format(self->expr, ignore_case);
+  if (!res)
+    return FALSE;
+
+  *dst = res->str;
+  *len = res->len;
+  return TRUE;
+}
+
+static gboolean
+_cache_needle(guint64 index, FilterXExpr *value, gpointer user_data)
+{
+  gboolean *ignore_case = ((gpointer *)user_data)[0];
+  GPtrArray *cached_strings = ((gpointer *) user_data)[1];
+
+  FilterXStringWithCache *obj_with_cache =  _string_with_cache_new(value, *ignore_case);
+  if(!obj_with_cache)
+    return FALSE;
+
+  g_ptr_array_add(cached_strings, obj_with_cache);
+  return TRUE;
+}
+
+
+static gboolean
+_expr_affix_init_needle(FilterXExprAffix *self, FilterXExpr *needle)
+{
+  self->needle.expr = filterx_expr_ref(needle);
+  if (filterx_expr_is_literal(self->needle.expr))
+    {
+      FilterXStringWithCache *obj_with_cache =  _string_with_cache_new(self->needle.expr, self->ignore_case);
+      if (!obj_with_cache)
+        return FALSE;
+      g_ptr_array_add(self->needle.cached_strings, obj_with_cache);
+    }
+  if (filterx_expr_is_literal_list_generator(self->needle.expr))
+    {
+      gpointer user_data[] = {&self->ignore_case, self->needle.cached_strings};
+      if(!filterx_literal_list_generator_foreach(self->needle.expr, _cache_needle, user_data))
+        return FALSE;
+    }
+  return TRUE;
+}
+
+static void
+_expr_affix_free(FilterXExpr *s)
+{
+  FilterXExprAffix *self = (FilterXExprAffix *) s;
+  g_ptr_array_free(self->needle.cached_strings, TRUE);
+
+  filterx_function_free_method(&self->super);
+}
+
+static GPtrArray *
+_expr_affix_eval_needle(FilterXExprAffix *self)
+{
+  FilterXObject *needle_obj = filterx_expr_eval(self->needle.expr);
+  if (!needle_obj)
+    return NULL;
+  GPtrArray *needles = NULL;
+
+  if (filterx_object_is_type(needle_obj, &FILTERX_TYPE_NAME(string)))
```

Comment:
`filterx_expr_eval()` might return a `message_value` of `LM_VT_STRING`, which won't pass this condition, but should.

When you fix this, please extend the E2E testcase to cover this.

Reply:
This is still not fixed and tested properly. This config reproduces the problem:

```
log {
  source { example-msg-generator(num(1) template("fooBARbAz") values("msg_val" => "foo")); };
  filterx {
    result = json();
    var = $msg_val;
    result.startswith_foo = startswith($MSG, var);
  };
  destination { stdout(); };
};
```

Reply:
Thanks, this ought to take care of it now:
```
   log_msg_value = ${values.str};
    if (includes(log_msg_value, log_msg_value))
    {
        result.works_with_message_value = true;
    };
```